mlfinlab/portfolio_optimization/cla.py

```diff
@@ -0,0 +1,608 @@
+        :param mean_f: (np.array) expected returns of free assets
+        :param w_b: (np.array) bounded asset weight values
+        :param asset_index: (int) index of the asset in the portfolio
+        :param b_i: (list) list of upper and lower bounded weight values
+        :return: (float) lambda value
+        '''
+
+        # Compute C
+        ones_f = np.ones(mean_f.shape)
+        c_1 = np.dot(np.dot(ones_f.T, covar_f_inv), ones_f)
+        c_2 = np.dot(covar_f_inv, mean_f)
+        c_3 = np.dot(np.dot(ones_f.T, covar_f_inv), mean_f)
+        c_4 = np.dot(covar_f_inv, ones_f)
+        c_final = -1*c_1 * c_2[asset_index] + c_3 * c_4[asset_index]
+        if c_final == 0:
+            return None, None
+
+        # Compute bi
+        if isinstance(b_i, list):
+            b_i = self._compute_bi(c_final, b_i)
+
+        # Compute Lambda
+        if w_b is None:
+
+            # All free assets
+            return float((c_4[asset_index] - c_1 * b_i) / c_final), b_i
+
+        ones_b = np.ones(w_b.shape)
+        l_1 = np.dot(ones_b.T, w_b)
+        l_2 = np.dot(covar_f_inv, covar_fb)
+        l_3 = np.dot(l_2, w_b)
+        l_2 = np.dot(ones_f.T, l_3)
+        lambda_value = float(((1 - l_1 + l_2) * c_4[asset_index] - c_1 * (b_i + l_3[asset_index])) / c_final)
+        return lambda_value, b_i
+
+    def _get_matrices(self, free_weights):
+        '''
+        Calculate the required matrices between free and bounded assets
+
+        :param free_weights: (list) list of free assets/weights
+        :return: (tuple of np.array matrices) the corresponding matrices
+        '''
+
+        covar_f = self._reduce_matrix(self.cov_matrix, free_weights, free_weights)
+        mean_f = self._reduce_matrix(self.expected_returns, free_weights, [0])
+        bounded_weights = self._get_bounded_weights(free_weights)
+        covar_fb = self._reduce_matrix(self.cov_matrix, free_weights, bounded_weights)
+        w_b = self._reduce_matrix(self.weights[-1], bounded_weights, [0])
+        return covar_f, covar_fb, mean_f, w_b
+
+    def _get_bounded_weights(self, free_weights):
+        '''
+        Compute the list of bounded assets
+
+        :param free_weights: (np.array) list of free weights/assets
+        :return: (np.array) list of bounded assets/weights
+        '''
+
+        return self._diff_lists(list(range(self.expected_returns.shape[0])), free_weights)
+
+    @staticmethod
+    def _diff_lists(list_1, list_2):
+        '''
+        Calculate the set difference between two lists
+        '''
+
+        return list(set(list_1) - set(list_2))
+
+    @staticmethod
+    def _reduce_matrix(matrix, row_indices, col_indices):
+        '''
+        Reduce a matrix to the provided set of rows and columns
+        '''
+
+        return matrix[np.ix_(row_indices, col_indices)]
+
+    def _purge_num_err(self, tol):
+        '''
+        Purge violations of inequality constraints (associated with ill-conditioned cov matrix)
```

Comment:
Docstrings... Please always add docs for arguments.

Reply:
Resolved